File: 0051 - N-Queens/solution.cpp
class Solution {
private:
    vector<bool> col, dia1, dia2;
    void place(int i, int n, vector<string>& now, vector<vector<string>>& res){
        if(i == n) {res.push_back(now); return;}
        for(int j = 0; j < n; j++){
            if(col[j] || dia1[i+j] || dia2[i-j+(n-1)]) continue;
            col[j] = dia1[i+j] = dia2[i-j+(n-1)] = true;
            now[i][j] = 'Q';
            place(i+1, n, now, res);
            now[i][j] = '.';
            col[j] = dia1[i+j] = dia2[i-j+(n-1)] = false;
        }
    }
public:
    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> res;
        vector<string> now(n, string(n, '.'));
        col.resize(n), dia1.resize(n*2-1), dia2.resize(n*2-1);
        place(0, n, now, res);
        return res;
    }
};

File: 0051 - N-Queens/solution.py
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        def place(i):
            if i == n:
                res.append([''.join(s) for s in now])
                return
            for j in range(n):
                if not (col[j] or dia1[i+j] or dia2[i-j+n-1]):
                    col[j] = dia1[i+j] = dia2[i-j+n-1] = True
                    now[i][j] = 'Q'
                    place(i+1)
                    now[i][j] = '.'
                    col[j] = dia1[i+j] = dia2[i-j+n-1] = False
        res, now = [], [['.']*n for _ in range(n)]
        col, dia1, dia2 = [False]*n, [False]*(n*2-1), [False]*(n*2-1)
        place(0)
        return res